site/content/blog/2021-04-23-optimizing-go-compiler-1.md

```diff
@@ -0,0 +1,227 @@
+
+So that was it. I had then committed to fix an issue I knew nothing about. Of course, I started regretting that decision the very moment after I hit Send. I know now that it was a great decision, but at that time it looked quite daunting.
+
+# The issue
+
+Ok, so let's dive into the issue itself. The bug in the compiler can be explained with the following function:
+
+```go
+import "encoding/binary"
+
+func f(b []byte, x *[8]byte) {
+	_ = b[8]
+	t := binary.LittleEndian.Uint64(x[:])
+	binary.LittleEndian.PutUint64(b, t)
+}
+```
+
+The report said that this function compiled to a series of small `MOVx`, but that it could be optimized to use instead a couple of `MOVQ`: one for loading and one for storing.
+
+If this sounds as gibberish, that's ok. Bear with me and we will get to what that sentence means. First, let's see what the function is supposed to do. To make things a bit easier, let's simply rewrite the variable names and add a couple of comments:
+
+```go
+import "encoding/binary"
+
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
+}
+```
+
+It is now more clear what the function does. It stores the contents of the second argument, the `src` array, into the first argument, the slice `dest`. It uses the [encoding/binary](https://golang.org/pkg/encoding/binary/) package for reading and writing the 8 bytes as an unsigned integer of 64 bits. For doing the copy, we use a temporal variable, `temp`.
+
+The only weird line might be the first one, which is a bounds check. This forces the compiler to check that `dest[8]`, and, thus, all `dest[0]`, ..., `dest[7]`, are valid indices. We do that at the very beginning, so the compiler knows that all indices we are using in the copy are valid, for the whole scope of the function. We do not need to do such a check with `src`, because its length is known at compile time: 8 bytes.
+
+So what's the bug? This is where things start to get interesting. The compiler is generating correct code, but it is not as optimized as it can be.
+
+# Understanding the compiler
+
+Before diving into the bug, let's pause for a bit to learn something more about how the Go compiler works.
+
+For anyone who does not know, a compiler is a translator: it converts a piece of code in a language (in our case, Go) to an equivalent piece of code in another language (in our case, assembly code, whose specific instructions depend on the specific architecture we want to run the executable in). Hopefully, the meaning of what we wrote in Go (the behavior of our program), is maintained throughout the process of translation.
```

Comment:
You are already writing the text here, so I'd perhaps avoid saying "For anyone who does not know". It might make a beginner feel bad, that they "should" know what a compiler is. Perhaps better to just write "Essentially, a compiler is .."

Reply:
Completely true, thanks for the suggestion!